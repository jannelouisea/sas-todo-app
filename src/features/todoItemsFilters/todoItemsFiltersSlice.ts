import { createSlice, PayloadAction } from '@reduxjs/toolkit'

import { Sort } from 'src/static/enums'
import { ITodoItemsFilters } from 'src/interfaces';

const initialState: ITodoItemsFilters = {
    searchTerm: '',
    showNewer: true,
    showOlder: true,
    showCompleted: true,
    sortBy: Sort.Incmpl1st,
};

const todoItemsFiltersSlice = createSlice({
    name: 'todoItemsFilters',
    initialState,
    reducers: {
        searchTermChanged(state, action: PayloadAction<string>) {
            return { ...state, searchTerm: action.payload }
        },
        showNewerToggled(state) {
            return { ...state, showNewer: !state.showNewer }
        },
        showOlderToggled(state) {
            return { ...state, showOlder: !state.showOlder }
        },
        showCompletedToggled(state) {
            return { ...state, showCompleted: !state.showCompleted }
        },
        sortByChanged(state, action: PayloadAction<Sort>) {
            return { ...state, sortBy: action.payload }
        }
    }
});

export const {
    searchTermChanged,
    showNewerToggled,
    showOlderToggled,
    showCompletedToggled,
    sortByChanged,
} = todoItemsFiltersSlice.actions;

export default todoItemsFiltersSlice.reducer;