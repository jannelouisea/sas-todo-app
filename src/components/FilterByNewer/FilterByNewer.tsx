import {
    FilterByButton
} from 'src/components'
import HourglassTopIcon from '@mui/icons-material/HourglassTop';
import PropTypes from 'prop-types';
import {
    MUIColor,
    MUIFontSize
} from 'src/static/enums'

import { useAppDispatch, useAppSelector } from 'src/app/hooks';
import { showNewerToggled } from 'src/features/todoItemsFilters/todoItemsFiltersSlice';

type Props = {
    className: string
}

function FilterByNewer({ className }: Props) {
    const showNewer = useAppSelector(state => state.todoItemsFilters.showNewer);
    const dispatch = useAppDispatch();

    const onToggleFilter = () => {
        dispatch(showNewerToggled());
    }

    return <FilterByButton
        className={className}
        label='Newer'
        color={MUIColor.Primary}
        startIcon={<HourglassTopIcon fontSize={MUIFontSize.Small} />}
        filterEnabled={showNewer}
        onClick={onToggleFilter}
        ariaLabel='Filter by newer'
    />;
}

FilterByNewer.propTypes = {
    className: PropTypes.string,
}

FilterByNewer.defaultProps = {
    className: ''
}

export default FilterByNewer;