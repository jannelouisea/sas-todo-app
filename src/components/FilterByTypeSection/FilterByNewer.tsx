import {
    FilterByButton
} from 'src/components'
import HourglassTopIcon from '@mui/icons-material/HourglassTop';
import PropTypes from 'prop-types';

import { useAppDispatch, useAppSelector } from 'src/app/hooks';
import { showNewerToggled } from 'src/features/todoItemsFilters/todoItemsFiltersSlice';

type Props = {
    className: string
}

function FilterByNewer({ className }: Props) {
    const showNewer = useAppSelector(state => state.todoItemsFilters.showNewer);
    const dispatch = useAppDispatch();

    const onToggleFilter = () => {
        dispatch(showNewerToggled());
    }

    return <FilterByButton
        className={className}
        label='Newer'
        color='primary'
        startIcon={<HourglassTopIcon fontSize='small' />}
        filterEnabled={showNewer}
        onClick={onToggleFilter}
    />;
}

FilterByNewer.propTypes = {
    className: PropTypes.string,
}

FilterByNewer.defaultProps = {
    className: ''
}

export default FilterByNewer;