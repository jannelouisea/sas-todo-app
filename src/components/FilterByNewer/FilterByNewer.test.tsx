import FilterByNewer from './FilterByNewer';
import { Provider } from 'react-redux';
import {
    IStoreType,
    createStore
} from 'src/app/store';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import * as todoItemsFiltersActions from 'src/features/todoItemsFilters/todoItemsFiltersSlice';

describe('FilterByNewer', () => {
    let store: IStoreType;

    beforeEach(() => {
        store = createStore();
    });

    it('renders correctly', () => {
        const tree = renderer
            .create(
                <Provider store={store}>
                    <FilterByNewer className='foo' />
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('toggles newer filter on click', () => {
        const todoItemNewerSpy = jest.spyOn(todoItemsFiltersActions, 'showNewerToggled');

        render(
            <Provider store={store}>
                <FilterByNewer className='foo' />
            </Provider>
        );

        let button: HTMLInputElement = screen.getByLabelText('Filter by newer');
        expect(button).toBeInTheDocument();
        fireEvent.click(button);

        expect(todoItemNewerSpy).toHaveBeenCalledTimes(1);
    })
});