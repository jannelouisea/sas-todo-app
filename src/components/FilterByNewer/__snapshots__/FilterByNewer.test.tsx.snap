// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterByNewer renders correctly 1`] = `
<div
  className="foo"
>
  <button
    aria-label="Filter by newer"
    className="MuiButton-root MuiButton-variantSolid MuiButton-colorPrimary MuiButton-sizeSm foo css-1d374uw-JoyButton-root"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
  >
    <span
      className="MuiButton-startDecorator css-187ulk2-JoyButton-startDecorator"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
        data-testid="HourglassTopIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m6 2 .01 6L10 12l-3.99 4.01L6 22h12v-6l-4-4 4-3.99V2H6zm10 14.5V20H8v-3.5l4-4 4 4z"
        />
      </svg>
    </span>
    Newer
  </button>
</div>
`;
